src/SfxWeb/src/app/views/cluster/federation/federation.component.ts

```diff
@@ -0,0 +1,656 @@
+          else {
+            nodesMap.clear();
+            nameToIdMap.clear();
+            // filling of the hashmaps with collected data
+            let FMMNodeId = this.handleNodeEvents();
+
+            // setting up the FMM info
+            if (FMMNodeId != "") {
+              let node = nodesMap.get(FMMNodeId);
+              node.FMM = true;
+            }
+
+            // processing of FM events
+            this.eventStoreHandlerFMSubscription = this.eventStoreHandlerFM.eventsList.refresh().subscribe(success => {
+              if (!success) this.failedToLoadEvents = true;
+              else {
+                // getting information about primary replica for FM service
+                this.handleFMEvents();
+
+                // forming data for graph in ascending order of node ids
+                let mapArray = [...nodesMap].sort((a, b) => a[1].id.localeCompare(b[1].id));
+                let len = mapArray.length;
+                nodes = []; edges = [];
+                //let nodes: { id?: string; label?: string; title?: string; x?: number; y?: number; }[] = [];
+                //let edges: { id?: number; from?: string; to?: string; color?: string; }[] = [{ id: 1 }];
+                let lUpCnt = 0, lDownCnt = 0, lGrayCnt = 0;
+                for (let elem of mapArray) {
+                  let title: string = formTitle(elem);
+                  nodes.push({ id: elem[1].id, label: " " + elem[1].name + " ", title: title, color: getNodeColor(elem[1]) });
+                  let neighborhood = [...elem[1].leaseStates];
+                  for (let nElem of neighborhood)
+                    edges.push({ from: elem[1].id, to: nElem[0] });
+                  if (elem[1].status == "Up") ++lUpCnt;
+                  else if (elem[1].status == "Gray") ++lGrayCnt;
+                  else lDownCnt++;
+                }
+                drawGraph("full");
+                let lUp = document.getElementById('labUp'); lUp.innerText = "Number of UP nodes: " + lUpCnt;
+                let lDown = document.getElementById('labDown'); lDown.innerText = "Number of DOWN nodes: " + lDownCnt;
+                let lGray = document.getElementById('labGray'); lGray.innerText = "Number of GRAY nodes: " + lGrayCnt;
+              }
+            });
+            
+          }
+        });
+      }
+    });
+  }
+
+  // method being called when customer clicks Refresh button in SFX
+  // since Refresh rate can be set to some predefine time (ie, 5 seconds)
+  // this is probably place we don't want to call logics from, especially for big clusters
+  // because that would take a lot oftime (since we are looping through all of them)
+  // therefore, logics implemented should be removed from here and put into ngOnInit() inside subscription
+  // reason for staying in here currently is because of testing needs (timeline bar is not working in mock regime)
+  refresh(messageHandler?: IResponseMessageHandler): Observable<any> {
```

Comment:
refresh should be unnecessary - the expectation should be that the window will only load what is visible within that time range and only query new data when the slider/dates are changed.

Reply:
yes absolutely, I just used it for easier testing, but it's really unnecessary